internal/app/orchestrator/orchestrator.go

```diff
@@ -90,16 +105,172 @@ func New(ctx context.Context, l log.Logger, mapper mapper.Mapper, upstreamClient
+		// Mimic the aggregated key.
+		// TODO (insert issue) This key needs to be made more granular to
+		//      uniquely identify a request. User-specified defaults?
+		aggregatedKey = fmt.Sprintf("%s%s_%s", unaggregatedPrefix, req.GetNode().GetId(), req.GetTypeUrl())
+	}
+
+	// Register the watch for future responses.
+	err = o.cache.AddRequest(aggregatedKey, req)
+	if err != nil {
+		// If we fail to register the watch, we need to kill this stream by
+		// closing the response channel.
+		o.logger.With("err", err).With("key", aggregatedKey).With(
+			"req", req).Error(ctx, "failed to add watch")
+		close(o.downstreamResponseChannels[&req])
+		// Clean up.
+		o.downstreamResponseChannelsMu.Lock()
+		delete(o.downstreamResponseChannels, &req)
+		o.downstreamResponseChannelsMu.Unlock()
+		return nil, nil
+	}
+
+	// Check if we have a cached response first.
+	cached, err := o.cache.Fetch(aggregatedKey)
+	if err != nil {
+		// Log, and continue to propagate the response upstream.
+		o.logger.With("err", err).With("key", aggregatedKey).Error(ctx, "failed to fetch aggregated key")
+	}
+
+	if cached != nil && cached.Resp != nil {
+		// If we have a cached response, immediately return the result.
+		o.downstreamResponseChannels[&req] <- gcp.Response{
+			Request:            req,
+			Version:            cached.Resp.Raw.GetVersionInfo(),
+			ResourceMarshaled:  true,
+			MarshaledResources: cached.Resp.MarshaledResources,
+		}
+	} else {
+		// Otherwise, check if we have a upstream stream open for this
+		// aggregated key. If not, open a stream with the representative
+		// request.
+		//
+		// Locking is necessary here so that a simultaneous downstream request
+		// that maps to the same aggregated key doesn't result in two upstream
+		// streams.
+		o.upstreamResponseChannelseMu.Lock()
+		if o.upstreamResponseChannels[aggregatedKey] == nil {
+			upstreamResponseChan := o.upstreamClient.OpenStream(ctx, &req)
+			// Spin up a go routine to watch for upstream responses.
+			// One routine is opened per aggregate key.
+			go o.watchUpstream(ctx, aggregatedKey, upstreamResponseChan)
+			o.upstreamResponseChannels[aggregatedKey] = upstreamResponseChan
+		}
+		o.upstreamResponseChannelseMu.Unlock()
+	}
+
+	return o.downstreamResponseChannels[&req], nil
 }
 
 // Fetch implements the polling method of the config cache using a non-empty request.
-func (c *orchestrator) Fetch(context.Context, discovery.DiscoveryRequest) (*gcp.Response, error) {
+func (o *orchestrator) Fetch(context.Context, discovery.DiscoveryRequest) (*gcp.Response, error) {
 	return nil, fmt.Errorf("Not implemented")
 }
 
-func (c *orchestrator) onCacheEvicted(key string, resource cache.Resource) {
-	// TODO implement.
+// watchResponse is intended to be called in a goroutine, to receive incoming
+// responses and fan out to downstream clients.
+func (o *orchestrator) watchUpstream(
+	ctx context.Context,
+	aggregatedKey string,
+	responseChannel <-chan *upstream.Response,
+) {
+	for {
```

Comment:
This needs a shutdown handler. One option is to listen for a shutdown channel and break out of the for loop.

Reply:
done, https://github.com/envoyproxy/xds-relay/pull/53/files#diff-2958638164db92742469d2bb08d0001aR232